File: customer/test/home/banner_link_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:shopxy_customer/features/home/domain/banner_link.dart';

void main() {
  group('BannerLink.parse', () {
    test('parses each supported kind', () {
      expect(
        BannerLink.parse('product:sp2rhACi'),
        const BannerLink(BannerLinkKind.product, 'sp2rhACi'),
      );
      expect(
        BannerLink.parse('category:home-kitchen'),
        const BannerLink(BannerLinkKind.category, 'home-kitchen'),
      );
      expect(
        BannerLink.parse('shop:sharma-electronics'),
        const BannerLink(BannerLinkKind.shop, 'sharma-electronics'),
      );
      expect(
        BannerLink.parse('search:winter jackets'),
        const BannerLink(BannerLinkKind.search, 'winter jackets'),
      );
    });

    test('tolerates whitespace and kind casing', () {
      expect(
        BannerLink.parse('  Category: Home-Kitchen  '),
        const BannerLink(BannerLinkKind.category, 'home-kitchen'),
      );
    });

    test('keeps a search phrase verbatim', () {
      expect(
        BannerLink.parse('search:Winter Jackets'),
        const BannerLink(BannerLinkKind.search, 'Winter Jackets'),
      );
    });

    test('returns null for the legacy formats rather than guessing', () {
      expect(BannerLink.parse('https://example.com/sale'), isNull);
      expect(BannerLink.parse('/shop/acme'), isNull);
      expect(BannerLink.parse('url:https://example.com'), isNull);
    });

    test('rejects unknown kinds and malformed values', () {
      expect(BannerLink.parse('collection:summer'), isNull);
      expect(BannerLink.parse('category:Home Kitchen'), isNull);
      expect(BannerLink.parse('category:home_kitchen'), isNull);
      expect(BannerLink.parse('shop:-leading-hyphen'), isNull);
      expect(BannerLink.parse('product:has spaces'), isNull);
    });

    test('rejects empty, missing and separator-less input', () {
      expect(BannerLink.parse(null), isNull);
      expect(BannerLink.parse(''), isNull);
      expect(BannerLink.parse('   '), isNull);
      expect(BannerLink.parse('product:'), isNull);
      expect(BannerLink.parse('product'), isNull);
      expect(BannerLink.parse(':value'), isNull);
    });

    test('caps a search phrase', () {
      expect(BannerLink.parse('search:${'a' * 120}'), isNotNull);
      expect(BannerLink.parse('search:${'a' * 121}'), isNull);
    });

    test('round-trips through toString', () {
      for (final raw in [
        'product:sp2rhACi',
        'category:home-kitchen',
        'shop:sharma-electronics',
        'search:winter jackets',
      ]) {
        final parsed = BannerLink.parse(raw)!;
        expect(parsed.toString(), raw);
        expect(BannerLink.parse(parsed.toString()), parsed);
      }
    });
  });
}

File: backend/tests/banners/banner-link.test.ts
import { describe, it, expect } from 'vitest';
import {
  parseBannerLink,
  formatBannerLink,
} from '../../src/modules/banners/banner-link.js';

describe('parseBannerLink', () => {
  it('parses each supported kind', () => {
    expect(parseBannerLink('product:sp2rhACi')).toEqual({
      kind: 'product',
      value: 'sp2rhACi',
    });
    expect(parseBannerLink('category:home-kitchen')).toEqual({
      kind: 'category',
      value: 'home-kitchen',
    });
    expect(parseBannerLink('shop:sharma-electronics')).toEqual({
      kind: 'shop',
      value: 'sharma-electronics',
    });
    expect(parseBannerLink('search:winter jackets')).toEqual({
      kind: 'search',
      value: 'winter jackets',
    });
  });

  it('tolerates surrounding whitespace and kind casing', () => {
    expect(parseBannerLink('  Category: Home-Kitchen  ')).toEqual({
      kind: 'category',
      value: 'home-kitchen',
    });
  });

  it('keeps a search phrase verbatim, including its case and spaces', () => {
    expect(parseBannerLink('search:Winter Jackets')).toEqual({
      kind: 'search',
      value: 'Winter Jackets',
    });
  });

  it('returns null for the legacy formats, rather than guessing', () => {
    expect(parseBannerLink('https://example.com/sale')).toBeNull();
    expect(parseBannerLink('/shop/acme')).toBeNull();
    expect(parseBannerLink('url:https://example.com')).toBeNull();
  });

  it('rejects an unknown kind', () => {
    expect(parseBannerLink('collection:summer')).toBeNull();
    expect(parseBannerLink('promo:xyz')).toBeNull();
  });

  it('rejects a malformed value for its kind', () => {
    expect(parseBannerLink('category:Home Kitchen')).toBeNull();
    expect(parseBannerLink('category:home_kitchen')).toBeNull();
    expect(parseBannerLink('shop:-leading-hyphen')).toBeNull();
    expect(parseBannerLink('product:has spaces')).toBeNull();
  });

  it('rejects empty, missing and separator-less input', () => {
    expect(parseBannerLink(null)).toBeNull();
    expect(parseBannerLink(undefined)).toBeNull();
    expect(parseBannerLink('')).toBeNull();
    expect(parseBannerLink('   ')).toBeNull();
    expect(parseBannerLink('product:')).toBeNull();
    expect(parseBannerLink('product')).toBeNull();
    expect(parseBannerLink(':value')).toBeNull();
  });

  it('caps a search phrase so a link cannot carry a payload', () => {
    expect(parseBannerLink(`search:${'a'.repeat(120)}`)).not.toBeNull();
    expect(parseBannerLink(`search:${'a'.repeat(121)}`)).toBeNull();
  });

  it('round-trips through formatBannerLink', () => {
    for (const raw of [
      'product:sp2rhACi',
      'category:home-kitchen',
      'shop:sharma-electronics',
      'search:winter jackets',
    ]) {
      const parsed = parseBannerLink(raw)!;
      expect(formatBannerLink(parsed)).toBe(raw);
      expect(parseBannerLink(formatBannerLink(parsed))).toEqual(parsed);
    }
  });
});
